### ISSUE TITLE
Problem parsing empty initial state

### ISSUE THREAD

**Issue by Maumagnaguagno** [LAPKT-dev/LAPKT-public#41]
Hello everyone,
I was thinking about removing the empty initial states from my parser testing problems, and decided to test the instance on the planning.domains editor (which I just discovered uses LAPKT) before commiting my changes (just to be sure it is compatible).

It seems that ``(:init)`` is not good, with a syntax error, and removing the empty initial state line is also unsupported, with ``Failed to parse the problem -- 'No subtree with name :init found in this tree'``.
At least one of those must be supported for problems with an empty initial state to be executed.
You can obtain my [domain](https://github.com/Maumagnaguagno/HyperTensioN/blob/master/examples/basic/basic.pddl) and [problem](https://github.com/Maumagnaguagno/HyperTensioN/blob/master/examples/basic/pb1.pddl) online.

I believe at least ``(:init)`` should be supported, not sure about the implicit empty initial state case.

**Comment by miquelramirez** [comment]
Hello @Maumagnaguagno,

What parser is being used? I am pretty sure though that omitting ```(:init )``` will trip parsers that combine domain and instance into one single token stream, even if an ```(:init )``` section may not be required by the grammar.

What syntax error are you getting when having an empty (all atoms false) ```(:init )``` condition?

**Comment by haz** [comment]
I think it's the FF parser. @nirlipo ?

**Comment by Maumagnaguagno** [comment]
I am not sure which parser is the being used by planning.domains, but I changed no configurations and loaded no plugins, so it is the default one.
The implicit empty goal gives a more verbose output that may help you track down the error:
```
Failed to parse the problem -- 'No subtree with name :init found in this tree'

/app/plan: line 3: 23255 Segmentation fault      timeout 10 "$(dirname "$0")"/siw-then-bfsf --domain $1 --problem $2 --output $3
```

**Comment by miquelramirez** [comment]
> This may trip parsers as you say, but **should be easy to fix**.
You start with an empty state structure and add elements contained in :init, otherwise you risk having a null structure if the initial state is not described, the same is valid for goal state.

Sure, feel free to send a PR with the fix.

**Comment by Maumagnaguagno** [comment]
Sorry about the ``easy to fix`` comment, I didn’t mean to disrespect, this is just the solution I currently use.
I am not familiar enough with your codebase to help with code, I just wanted to help with bug tracking.

**Comment by miquelramirez** [comment]
Hi @Maumagnaguagno ,

no offense was taken or apology is needed - I sincerely thought that you had a plan for this.

I will let @nirlipo pronounce himself, as the maintainer of LAPKT, but `ff` is not really part of the codebase, rather being a third-party module that we retrofitted 10 years ago to provide basic parsing and grounding for "IPC-ready" planners. As you are aware, `ff` is by now almost old enough to vote and have political rights in most Western countries. Issues like this one, which I suspect have to do with deeply ingrained design decisions which probably is already 15 years too late to review, will have to get worked around.

In the specific case of this issue I wonder if it is possible to "flip inside out" your problem formulation, so the ground atoms correspond to the negative literals of the atoms of your original formulation. That way, what was the literal (expressed in the ADL fragment of PDDL)

```
(not (foo a b)) 
```

would become

```
(not_foo a b)
```

Such a transformation may or not be possible.

Alternatively, it could be investigated whether the front-end based on `Fast Downward`'s `translate` Python module could be used instead. I am not sure whether that would provide better results, as similar assumptions may be built-in into the parsing, grounding and invariant calculation procedures.

Regards,

Miquel.

**Comment by haz** [comment]
Ya, shifting to FD is a no-go for timing reasons. Online planner needs to be agile to the max.

Another option is to have a dummy fluent -- define the predicate with 0-arity, include it in the initial state, and leave it at that. Not pretty, but likely enough to shush the parser.

**Comment by Maumagnaguagno** [comment]
You are right @miquelramirez, flipping ground atoms may be impossible, as one needs to enumerate them all before adding to the flipped initial state.
I did exactly what @haz commented as a sanity check yesterday :smile:.
I would also add the dummy predicate in the goal if the goal is empty or missing, even with no goals being a really weird description.

**Comment by nirlipo** [comment]
That's the correct file @Maumagnaguagno, It defines the FF parser. @haz it's true that for the online planner the FF parser is generally faster, we should benchmark the clingo parser too at some point. I'll add it to my todo list :) 

Mau, It would be great If you can send the patch for the yacc file!

**Comment by Maumagnaguagno** [comment]
I am having trouble trying to compile within Manjaro.
Getting stuck on makedepend not being able to find standard libraries, probably missing some path, as GCC is able to find them.
I will try again later.

**Comment by nirlipo** [comment]
right, I have no experience with manjaro, FF parser other main dependencies are bison and flex, I'm not sure why some std libraries are missing.
